### PR TITLE
Rewriting C code for scan_object, get_size, and get_object_start_ref in Rust

### PR DESCRIPTION
**Pull request by udesou**
This PR refactors the code for scanning Julia objects in Rust and removes the C variant. It does the same for the functions `get_so_object_size` and `get_object_start_ref`. 
It also updates the types in `mmtk_julia_types.h`, from which we generate a new version of `julia_types.rs` using Bindgen. Since Bindgen generates inefficient code for accessing bitfields, I've also added "custom" functions to access those fields (only the ones used in `scan_object`, `get_so_object_size` and `get_object_start_ref`).  

NB: needs to be merged with https://github.com/mmtk/julia/pull/25

### CODE REVIEW

File: mmtk/src/julia_scanning.rs [hunk 1]
```diff
@@ -30,285 +36,359 @@ pub unsafe fn mmtk_jl_typeof(addr: Address) -> Address {
...
+        return ty;
+    }
+    return t.to_ptr::<mmtk_jl_datatype_t>();
+}
+
+const PRINT_OBJ_TYPE: bool = false;
+
+// INFO: *_custom() functions are acessors to bitfields that do not use bindgen generated code.
+#[inline(always)]
+pub unsafe fn scan_julia_object<EV: EdgeVisitor<JuliaVMEdge>>(obj: Address, closure: &mut EV) {
```
Comment (qinsoon) [inline mmtk/src/julia_scanning.rs:58]: Can you add a comment saying
> This function is a rewrite of `gc_mark_outrefs()` in `gc.c`. We need to keep this code updated if Julia updates their object layout and their scanning code.

File: mmtk/src/julia_scanning.rs [hunk 2]
```diff
@@ -344,14 +461,88 @@ pub fn process_offset_edge(
...
+
+#[inline(always)]
+pub unsafe fn mmtk_jl_tparam0(vt: *const mmtk_jl_datatype_t) -> *const mmtk_jl_datatype_t {
+    mmtk_jl_svecref((*vt).parameters, 0)
+}
+
+#[inline(always)]
+pub unsafe fn mmtk_jl_svecref(vt: *mut mmtk_jl_svec_t, i: usize) -> *const mmtk_jl_datatype_t {
+    // FIXME: missing debug (?) assertions
+    // assert(jl_typetagis(t,jl_simplevector_tag << 4));
```
Comment (qinsoon) [inline mmtk/src/julia_scanning.rs:524]: You can just add these assertions as `debug_assert`.

File: mmtk/src/julia_scanning.rs [hunk 3]
```diff
@@ -30,285 +36,359 @@ pub unsafe fn mmtk_jl_typeof(addr: Address) -> Address {
...
     }
 }
 
+unsafe fn mmtk_scan_gcstack<EV: EdgeVisitor<JuliaVMEdge>>(
+    ta: *const mmtk_jl_task_t,
+    closure: &mut EV,
+) {
+    let stkbuf = (*ta).stkbuf;
+    let copy_stack = (*ta).copy_stack_custom();
+    // FIXME: the code below is executed COPY_STACKS has been defined in the C Julia implementation - it is on by default
```
Comment (qinsoon) [inline mmtk/src/julia_scanning.rs:316]: What does this comment mean?

**Review by qinsoon — APPROVED** [review]
LGTM

File: mmtk/src/julia_scanning.rs (hunk 3 above)
Comment (udesou) [reply mmtk/src/julia_scanning.rs:316]: The C code has a `#ifdef COPY_STACKS` but I'm not sure how to do this in Rust (https://github.com/mmtk/julia/blob/f690aa3a5621bfa1d6a07f911818f203d3f8d650/src/gc.c#L2118).

It is set here in `options.h`: https://github.com/mmtk/julia/blob/f690aa3a5621bfa1d6a07f911818f203d3f8d650/src/options.h#L102

Comment (qinsoon) [reply mmtk/src/julia_scanning.rs:316]: Rust has `cfg` for conditional compilation. If necessary, you can turn the code into a block that is guarded with `#[cfg(feature = "julia_copy_stack")]`. As we build the Rust project separately from the C project, there is no way to get the flag from C in compile time. The only thing you can do is to have an assertion to make sure that `COPY_STACK` is defined in C.

Comment (udesou) [reply mmtk/src/julia_scanning.rs]: I'm not sure where to put the check. I've added [this function](https://github.com/udesou/mmtk-julia/blob/30f1e676b9e3d4768f99355f7368a28daee6269e/mmtk/src/util.rs#L100)  and I'm calling it before initialising MMTk (https://github.com/udesou/julia/blob/ed30d1c925865dd5c4ca2482701e20d6bc4700ec/src/mmtk-gc.c#L365). Would you suggest I do something different?

**Pull request merged**